lib/collection/src/shards/local_shard/clock_map.rs

```diff
@@ -0,0 +1,148 @@
+use std::collections::{hash_map, HashMap};
+use std::io::Write as _;
+use std::path::Path;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::{cmp, fs, io};
+
+use serde::{Deserialize, Serialize};
+
+use crate::operations::types::CollectionError;
+use crate::operations::ClockTag;
+use crate::shards::shard::PeerId;
+
+#[derive(Debug, Default, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct ClockMap {
+    clocks: HashMap<Key, Clock>,
+}
+
+impl ClockMap {
+    pub fn load_or_default(path: &Path) -> Result<Self> {
+        let result = Self::load(path);
+
+        if let Err(Error::Io(err)) = &result {
+            if err.kind() == io::ErrorKind::NotFound {
+                return Ok(Self::default());
+            }
+        }
+
+        result
+    }
+
+    pub fn load(path: &Path) -> Result<Self> {
+        let file = fs::File::open(path)?;
+        let clock_map = serde_json::from_reader(io::BufReader::new(file))?;
+        Ok(clock_map)
+    }
+
+    pub fn store(&self, path: &Path) -> Result<()> {
+        let file = fs::File::create(path)?;
+
+        let mut writer = io::BufWriter::new(file);
+        serde_json::to_writer(&mut writer, &self)?;
+        writer.flush()?;
+
+        Ok(())
+    }
+
+    pub fn advance_clock_and_correct_tag(&mut self, clock_tag: &mut ClockTag) -> bool {
+        let (clock_updated, current_tick) = self.advance_clock_impl(clock_tag);
+
+        // We "accept" an operation, if it has `clock_tick` that is "newer" than `current_tick` in `ClockMap`
+        // (e.g., if `advance_clock_impl` *updated* the clock and returned `clock_updated = true`).
+        //
+        // If we "reject" an operation (because it has `clock_tick` that is "older" than `current_tick` in `ClockMap`),
+        // we have to update its clock tag with `current_tick`, so that it can be "echoed" back to the node.
+        //
+        // And we also *always* accept all operations with `clock_tick = 0` and *always* update their clock tags.
+
+        let operation_accepted = clock_updated || clock_tag.clock_tick == 0;
+        let update_tag = !clock_updated || clock_tag.clock_tick == 0;
+
+        let (operation_accepted, update_tag) = if clock_tag.clock_tick > 0 {
+            (clock_updated, !clock_updated)
+        } else {
+            (true, true)
+        };
```

Comment:
> I prefer to put "main" case in the first branch, but clock_tick == 0 is a special case here, but it's also super unobvious if you put clock_tick > 0 into conditional in previous example

I guess that kinda resolves it for me (I don't like both 2nd and 3rd example, so I'd go with the first), but I still want to hear what you guys think.

Reply:
I think I like the first one best.

That is because we check and assign the two separately, rather than in a tuple. For me that is a bit easier to understand.

Reply:
Remove second example from the code.